cli/config/credentials/oauth_store.go

```diff
@@ -0,0 +1,206 @@
+package credentials
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"os"
+	"time"
+
+	"github.com/docker/cli/cli/config/types"
+	"github.com/docker/cli/cli/oauth"
+)
+
+// oauthStore wraps an existing store that transparently handles oauth
+// flows, managing authentication/token refresh and piggybacking off an
+// existing store for storage/retrieval.
+type oauthStore struct {
+	backingStore Store
+	manager      oauth.Manager
+}
+
+// NewOAuthStore creates a new oauthStore backed by the provided store.
+func NewOAuthStore(backingStore Store, manager oauth.Manager) Store {
+	return &oauthStore{
+		backingStore: backingStore,
+		manager:      manager,
+	}
+}
+
+const minimumTokenLifetime = 50 * time.Minute
+
+// see https://github.com/moby/buildkit/pull/5165#discussion_r1682531996
+const defaultRegistry = "https://index.docker.io/v1/"
+
+// Get retrieves the credentials from the backing store, refreshing the
+// access token if the stored credentials are valid for less than minimumTokenLifetime.
+// If the credentials being retrieved are not for the official registry, they are
+// returned as is. If the credentials retrieved do not parse as a token, they are
+// also returned as is.
+func (o *oauthStore) Get(serverAddress string) (types.AuthConfig, error) {
+	if serverAddress != defaultRegistry {
+		return o.backingStore.Get(serverAddress)
+	}
+
+	tokenRes := o.fetchFromBackingStore()
+	if tokenRes == nil {
+		return o.backingStore.Get(serverAddress)
+	}
+
+	// if the access token is valid for less than minimumTokenLifetime, refresh it
+	if tokenRes.RefreshToken != "" && tokenRes.Claims.Expiry.Time().Before(time.Now().Add(minimumTokenLifetime)) {
```

Comment:
Looks like `Expiry` is `*NumericData` so it can be `nil`.
Should we do a nil-check just in case so we don't panic?